src/os/bluestore/simple_bitmap.cc

```diff
@@ -0,0 +1,325 @@
+#include "bluestore_types.h"
+#include "common/debug.h"
+#define dout_context cct
+#define dout_subsys  ceph_subsys_bluestore
+
+#define dout_context cct
+#define dout_subsys ceph_subsys_bluestore
+#undef dout_prefix
+#define dout_prefix *_dout << __func__ << "::SBMAP::" << this << " "
+
+static struct extent_t null_extent = {0, 0};
+
+//----------------------------------------------------------------------------
+std::ostream& operator<<(std::ostream& out, const SimpleBitmap& bmap)
+{
+  for (unsigned i = 0; i < bmap.m_word_count; i++) {
+    if(bmap.m_arr[i] == 0) {
+      continue;
+    }
+
+    out << "[" << i << "]" << "0x" << std::setfill('0') << std::setw(2) << std::right << std::hex;
+
+    if (SimpleBitmap::BYTES_IN_WORD > 1) {
+      out << bmap.m_arr[i] << std::endl;
+    } else {
+      // hex printout doesn't work for uint8_t, must convert to higher integer
+      out << (unsigned) bmap.m_arr[i] << std::endl;
+    }
+  }
+  return out;
+}
+
+//----------------------------------------------------------------------------
+//throw bad_alloc
+SimpleBitmap::SimpleBitmap(CephContext *_cct, const std::string &_path, uint64_t num_bits) :cct(_cct), path(_path)
+{
+  m_num_bits   = num_bits;
+  m_word_count = bits_to_words(num_bits);
+  if (num_bits & BITS_IN_WORD_MASK) {
+    m_word_count++;
+  }
+  m_arr = new uint64_t [m_word_count];
+  clear_all();
+  dout(20) << "m_word_count  = " << m_word_count  << dendl;
+  dout(20) << "m_num_bits    = " << m_num_bits    << dendl;
+}
+
+//----------------------------------------------------------------------------
+SimpleBitmap::~SimpleBitmap()
+{
+  dout(5) << __func__ << dendl;
+  delete [] m_arr;
+}
+
+
+//----------------------------------------------------------------------------
+void SimpleBitmap::__set_single_bit_no_bounderies_check(uint64_t offset)
+{
+  uint64_t word_index = offset_to_index(offset);
+  uint64_t bit_offset = (offset & BITS_IN_WORD_MASK);
+  uint64_t set_mask   = 1ULL << bit_offset;
+  m_arr[word_index] |= set_mask;
+}
+
+//----------------------------------------------------------------------------
+bool SimpleBitmap::set(uint64_t offset, uint64_t length)
+{
+  dout(20) <<" [" << std::hex << offset << ", " << length << "]" << dendl;
+
+  if (offset + length > m_num_bits) {
+    derr << "offset + length = " << offset + length << " exceeds map size = " << m_num_bits << dendl;
+    return false;
+  }
+
+  // special case optimization
+  if (length == 1) {
+    __set_single_bit_no_bounderies_check(offset);
+    return true;
+  }
+
+  // find the first word index
+  uint64_t index         = offset_to_index(offset);
+  // handle the first word which might be incomplete
+  uint64_t first_bit_set = (offset & BITS_IN_WORD_MASK);
+  if (first_bit_set != 0) {
+    uint64_t   set_mask      = FULL_MASK << first_bit_set;
+    uint64_t   first_bit_clr = first_bit_set + length;
+    if (first_bit_clr <= BITS_IN_WORD) {
+      if (first_bit_clr < BITS_IN_WORD) {
```

Comment:
there is 1/BITS_IN_WORD chance that first_bit_clr is eactly BITS_IN_WORD.
Is it worth making a separate branch for it?
I guess it would be better if it just calculated clr_bits, blr_mask and updated set_mask.